crypto/bio/bio_meth.c

```diff
@@ -10,26 +10,24 @@
 #include "bio_local.h"
 #include "internal/thread_once.h"
 
-CRYPTO_RWLOCK *bio_type_lock = NULL;
+static CRYPTO_REF_COUNT bio_type_count;
 static CRYPTO_ONCE bio_type_init = CRYPTO_ONCE_STATIC_INIT;
 
 DEFINE_RUN_ONCE_STATIC(do_bio_type_init)
 {
-    bio_type_lock = CRYPTO_THREAD_lock_new();
-    return bio_type_lock != NULL;
+    return CRYPTO_NEW_REF(&bio_type_count, BIO_TYPE_START);
 }
 
 int BIO_get_new_index(void)
 {
-    static CRYPTO_REF_COUNT bio_count = BIO_TYPE_START;
     int newval;
 
     if (!RUN_ONCE(&bio_type_init, do_bio_type_init)) {
         /* Perhaps the error should be raised in do_bio_type_init()? */
         ERR_raise(ERR_LIB_BIO, ERR_R_CRYPTO_LIB);
         return -1;
     }
-    if (!CRYPTO_UP_REF(&bio_count, &newval, bio_type_lock))
+    if (!CRYPTO_UP_REF(&bio_type_count, &newval))
```

Comment:
This looks like an abuse of the reference counters that is most likely doing the wrong thing. I don't think this actually guarantees unique numbers, just that it will get increased. Please use CRYPTO_atomic_add().

Reply:
There is also no CRYPTO_FREE_REF call.

Reply:
It probably does guarantee unique numbers.

Reply:
Yes, it does guarantee unique numbers.  There was no free call in the original code.  It allocated the lock and never freed it.  Since this is a pure refactoring, I'd prefer to not address that here.  I've raised an issue: #21277

Reply:
The reason it guarantees a unique number is because the base atomic atomic_fetch_add_explicit() happens to return the old value before incrementing. If it had returned the new value, it would not have been unique. Since this code only calls the CRYPTO_UP_REF, and never the CRYPTO_DOWN_REF it happens to work.